torchmetrics/functional/image/sam.py

```diff
@@ -0,0 +1,115 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Tuple, Union
+
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.regression.cosine_similarity import cosine_similarity
+from torchmetrics.utilities.checks import _check_same_shape
+from torchmetrics.utilities.distributed import reduce
+
+
+def _sam_update(preds: Tensor, target: Tensor) -> Tuple[Tensor, Tensor]:
+    """Updates and returns variables required to compute Spectral Angle Mapper. Checks for same shape and type of
+    the input tensors.
+
+    Args:
+        preds: Predicted tensor
+        target: Ground truth tensor
+    """
+
+    if preds.dtype != target.dtype:
+        raise TypeError(
+            "Expected `preds` and `target` to have the same data type."
+            f" Got preds: {preds.dtype} and target: {target.dtype}."
+        )
+    _check_same_shape(preds, target)
+    if len(preds.shape) != 4:
+        raise ValueError(
+            "Expected `preds` and `target` to have BxCxHxW shape."
+            f" Got preds: {preds.shape} and target: {target.shape}."
+        )
+    return preds, target
+
+
+def _sam_compute(
+    preds: Tensor,
+    target: Tensor,
+    reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
+) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+    """Computes Spectral Angle Mapper.
+
+    Args:
+        preds: estimated image
+        target: ground truth image
+        reduction: a method to reduce metric score over labels.
+
+            - ``'elementwise_mean'``: takes the mean (default)
+            - ``'sum'``: takes the sum
+            - ``'none'``: no reduction will be applied
+
+    Example:
+        >>> preds = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(42))
+        >>> target = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(123))
+        >>> preds, target = _sam_update(preds, target)
+        >>> _sam_compute(preds, target)
+        tensor(0.7377)
+    """
+    B, C, H, W = preds.shape
+    preds = preds.reshape(B, C, H * W)
+    target = target.reshape(B, C, H * W)
+    sam_score = torch.clip(cosine_similarity(preds, target, reduction="none"), -1, 1).arccos()
+    return reduce(sam_score, reduction)
+
+
+def universal_spectral_angle_mapper(
```

Comment:
To be coherent with the class metric `SpectralAngleMapper`, it seems that `spectral_angle_mapper` would be a better name.

```suggestion
def spectral_angle_mapper(
```

Reply:
I agree. I will revise at your suggestion.